source/opt/constructs.h

```diff
@@ -0,0 +1,306 @@
+class Instruction {
+ public:
+  // Creates a default OpNop instruction.
+  Instruction() : opcode_(SpvOpNop), type_id_(0), result_id_(0) {}
+  // Creates an instruction with the given opcode |op| and no additional logical
+  // operands.
+  Instruction(SpvOp op) : opcode_(op), type_id_(0), result_id_(0) {}
+  // Creates an instruction using the given spv_parsed_instruction_t |inst|. All
+  // the data inside |inst| will be copied and owned in this instance. And keep
+  // record of line-related debug instructions |dbg_line| ahead of this
+  // instruction, if any.
+  Instruction(const spv_parsed_instruction_t& inst,
+              std::vector<Instruction>&& dbg_line = {});
+
+  SpvOp opcode() const { return opcode_; }
+  uint32_t type_id() const { return type_id_; }
+  uint32_t result_id() const { return result_id_; }
+  // Returns the vector of line-related debug instructions attached to this
+  // instruction and the caller can directly modify them.
+  std::vector<Instruction>& dbg_line_insts() { return dbg_line_insts_; }
+
+  // Gets the number of logical operands.
+  uint32_t NumOperands() const {
+    return static_cast<uint32_t>(operands_.size());
+  }
+  // Gets the number of SPIR-V words occupied by all logical operands.
+  uint32_t NumOperandWords() const {
+    return NumInOperandWords() + TypeResultIdCount();
+  }
+  // Gets the |index|-th logical operand.
+  const Operand& GetOperand(uint32_t index) const;
+  // Gets the |index|-th logical operand as a single SPIR-V word. This will only
+  // consider the first word in the logical operand, even which has more than
+  // one words.
+  uint32_t GetSingleWordOperand(uint32_t index) const;
+  // Sets the |index|-th operand's data to the given |data|.
+  void SetOperand(uint32_t index, std::vector<uint32_t>&& data);
+
+  // The following methods are similar to the above, but are for in operands.
+  uint32_t NumInOperands() const {
+    return static_cast<uint32_t>(operands_.size() - TypeResultIdCount());
+  }
+  uint32_t NumInOperandWords() const;
+  const Operand& GetInOperand(uint32_t index) const {
+    return GetOperand(index + TypeResultIdCount());
+  }
+  uint32_t GetSingleWordInOperand(uint32_t index) const {
+    return GetSingleWordOperand(index + TypeResultIdCount());
+  }
+
+  // Returns true if this instruction is OpNop.
+  bool IsNop() const {
+    return opcode_ == SpvOpNop && type_id_ == 0 && result_id_ == 0 &&
+           operands_.empty();
+  }
+  // Turns this instruction to OpNop. This does not clear out all preceding
+  // line-related debug instructions.
+  void ToNop() {
+    opcode_ = SpvOpNop;
+    type_id_ = result_id_ = 0;
+    operands_.clear();
+  }
+
+  // Runs the given function |f| on this instruction and preceding debug
+  // instructions.
+  void ForEachInst(const std::function<void(Instruction*)>& f);
+
+  // Pushes the binary segments for this instruction into the back of *|binary|.
+  // If |keep_nop| is true and this is a OpNop, do nothing.
```

Comment:
Good catch. Actually it took me a while to understand that this is indeed reversed. Changed.